src/app.ts

```diff
@@ -0,0 +1,295 @@
+	}
+
+	const handleConfError = () => { // don't let on if an email exists
+		handleError(req, res, 'EmailConfirmationError', `There was an error confirming ${email}`)
+	}
+
+	// Check if user exists
+	const { email, key } = validation.value;
+	const user = await users.findOne({email});
+	if (!user) {
+		handleConfError();
+		console.log( 'MissingUserError', `User ${email} does not exist`)
+	 	return;
+	}
+
+	// Check if user is already confirmed
+	if (!user.confirmation_key) {
+		handleConfError();
+		console.log( 'UserAlreadyConfirmedError', 'User is already confirmed');
+		return;
+	}
+
+	// Check if key is correct
+	if (key != user['confirmation_key']) {
+		handleConfError();
+		console.log('ConfirmationKeyError', `Key does not match`);
+		return;
+	}
+
+	// Update DB
+	await users.findOneAndUpdate({email: email}, { $unset: { confirmation_key: ''}});
+	console.log(`EMAIL_CONFIRMATION_SUCCESSFUL: with email [${email}]`);
+
+	// Create Stripe subscription
+	console.log('Awaiting subscription creation');
+	let subscription;
+	try {
+		subscription = await stripe.subscriptions.create({
+			customer: user.stripe_id,
+			items: [{ plan: 'plan_DrPVwslmSpiOT4' }]
+		});
+	} catch (err) {
+		res.send();
+		handleError(req, null, err.type, err.message);
+	 	return;
+	}
+
+	res.send();
+});
+
+app.post('/login', async function (req: Request, res: Response) {
+	const user = await users.findOne({email: req.body.email});
+	const isValid = sodium.crypto_pwhash_str_verify(user['pwhash'], req.body.password);
+	const email = req.body.email;
+	if (isValid) {
+		const dict = generateTokenDict(user['signing_key'].buffer);
+		console.log(`LOGIN_SUCCESSFUL: with username [${email}]`);
+		res.send( dict );
+	} else {
+		console.log(`LOGIN_FAILED: with username [${email}]`);
+		res.send( { error: 'Login error' });
+	}
+});
+
+const generateTokenDict = (key: Buffer) => {
+	const valid_until = (Date.now() + token_window).toString();
+	const mac = sodium.crypto_auth(valid_until, key);
+	return ({ mac: to_base64(mac), valid_until: valid_until});
+};
+
+const from_base64 = (bytes: string) => {
+	return sodium.from_base64(bytes, sodium.base64_variants.URLSAFE_NO_PADDING )
+}
+
+const to_base64 = (bytes: Uint8Array) => {
+	return sodium.to_base64(bytes, sodium.base64_variants.URLSAFE_NO_PADDING )
+}
+
+const handleError = (req: Request, res: Response, code: String, message: String) => {
+	const path = req.url.toUpperCase();
+	logMessage(req, `${ code }: ${ message }`);
+	if (res) {
+		res.status(400).json({ code, message });
+	}
+}
+
+const logMessage = (req: Request, message: String) => {
+	console.log(`${ req.url.toUpperCase() }: ${ message }` );
+}
+
+// function distillError(error: IStripeError) {
+// 	return (({ code, message, param, type }) => ({ code, message, param, type }))(error)
+// }
+
+function distillError(error: any) {
+	delete error.stack;
+	return error;
+}
+
+function close() {
+	server.close();
+	expressServer.close();
+}
+
+function getUsers () {
```

Comment:
this seems awkward to me. typically you might have a separate file in which you declare the users collection (and probably give it a schema - likely using mongoose even though i kind of hate mongoose), and then export the collection from that file. then this file could import `users`, as could your test files